src/stream/src/executor/now.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_stack_trace::StackTrace;
+use chrono::NaiveDateTime;
+use futures::{pin_mut, StreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::catalog::{Field, Schema};
+use risingwave_common::row::Row;
+use risingwave_common::types::{DataType, NaiveDateTimeWrapper, ScalarImpl};
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_common::util::epoch::Epoch;
+use risingwave_storage::StateStore;
+use tokio::sync::mpsc::UnboundedReceiver;
+
+use super::{
+    Barrier, BoxedMessageStream, Executor, Message, PkIndices, PkIndicesRef, StreamExecutorError,
+    Watermark,
+};
+use crate::common::table::state_table::StateTable;
+
+pub struct NowExecutor<S: StateStore> {
+    /// Receiver of barrier channel.
+    barrier_receiver: UnboundedReceiver<Barrier>,
+
+    pk_indices: PkIndices,
+    identity: String,
+    schema: Schema,
+    state_table: StateTable<S>,
+}
+
+impl<S: StateStore> NowExecutor<S> {
+    pub fn new(
+        barrier_receiver: UnboundedReceiver<Barrier>,
+        executor_id: u64,
+        state_table: StateTable<S>,
+    ) -> Self {
+        let schema = Schema::new(vec![Field {
+            data_type: DataType::Timestamp,
+            name: String::from("now"),
+            sub_fields: vec![],
+            type_name: String::default(),
+        }]);
+        Self {
+            barrier_receiver,
+            pk_indices: vec![],
+            identity: format!("NowExecutor {:X}", executor_id),
+            schema,
+            state_table,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn into_stream(self) {
+        let Self {
+            mut barrier_receiver,
+            mut state_table,
+            schema,
+            ..
+        } = self;
+
+        // Consume the first barrier message and initialize state table.
+        let barrier = barrier_receiver
+            .recv()
+            .stack_trace("now_executor_recv_first_barrier")
+            .await
+            .unwrap();
+        state_table.init_epoch(barrier.epoch);
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
```

Comment:
So we don't emit the timestamp on the first barrier? Is this expected?

Reply:
> So we don't emit the timestamp on the first barrier? Is this expected?

We can't because no message is permitted before the first barrier.

Reply:
> How about emitting the timestamp right after the barrier?

The timestamp will be updated on the second barrier so I think that it is unnecessary.